raft/log.go

```diff
@@ -514,29 +518,45 @@ func (l *Log) leader() (id uint64, u *url.URL) {
 // Join contacts a node in the cluster to request membership.
 // A log cannot join a cluster if it has already been initialized.
 func (l *Log) Join(u *url.URL) error {
-	l.mu.Lock()
-	defer l.mu.Unlock()
+	// Validate under lock.
+	var nodeURL *url.URL
+	if err := func() error {
+		l.mu.Lock()
+		defer l.mu.Unlock()
 
-	// Check if open.
-	if !l.opened() {
-		return ErrClosed
-	} else if l.id != 0 {
-		return ErrInitialized
-	} else if l.URL == nil {
-		return ErrURLRequired
+		if !l.opened() {
+			return ErrClosed
+		} else if l.id != 0 {
+			return ErrInitialized
+		} else if l.URL == nil {
+			return ErrURLRequired
+		}
+
+		nodeURL = l.URL
+		return nil
+	}(); err != nil {
+		return err
 	}
 
+	l.tracef("joining to: %s", u)
+
 	// Send join request.
-	id, config, err := l.transport().Join(u, l.URL)
+	id, config, err := l.Transport.Join(u, nodeURL)
 	if err != nil {
 		return err
 	}
 
+	l.tracef("confirmed join")
+
+	// Lock once the join request is returned.
+	l.mu.Lock()
```

Comment:
It's setting the state at the bottom. We also wouldn't want the `Log` to try to write the `id` or `config` files if someone called `Join()` twice. (not sure why anyone would though)